cmd/juju/application/bind.go

```diff
@@ -0,0 +1,238 @@
+	nArgs := len(args)
+	if nArgs == 0 {
+		return errors.Errorf("no application specified")
+	}
+
+	if !names.IsValidApplication(args[0]) {
+		return errors.Errorf("invalid application name %q", args[0])
+	}
+	c.ApplicationName = args[0]
+	c.BindExpression = strings.Join(args[1:], " ")
+	return nil
+}
+
+// Run connects to the specified environment and applies the requested binding
+// changes.
+func (c *bindCommand) Run(ctx *cmd.Context) error {
+	apiRoot, err := c.NewAPIRoot()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() { _ = apiRoot.Close() }()
+
+	if err = c.checkApplicationFacadeSupport(apiRoot, "changing application bindings", 11); err != nil {
+		return err
+	}
+
+	if err = c.parseBindExpression(apiRoot); err != nil {
+		return err
+	}
+
+	modelConfigGetter := c.NewModelConfigGetter(apiRoot)
+	modelConfig, err := getModelConfig(modelConfigGetter)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	generation, err := c.ActiveBranch()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	applicationClient := c.NewApplicationClient(apiRoot)
+	applicationInfo, err := applicationClient.Get(generation, c.ApplicationName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Validate endpoints and merge operator-defined bindings
+	curBindings := applicationInfo.EndpointBindings
+	var epList []string
+	for epName := range curBindings {
+		if epName == "" {
+			continue
+		}
+		epList = append(epList, epName)
+	}
+	curCharmEndpoints := set.NewStrings(epList...)
+
+	if err := c.validateEndpointNames(curCharmEndpoints, curBindings, c.Bindings); err != nil {
+		return errors.Trace(err)
+	}
+
+	appDefaultSpace := detectDefaultSpace(modelConfig, curBindings)
+
+	var bindingsChangelog []string
+	c.Bindings, bindingsChangelog = mergeBindings(curCharmEndpoints, curBindings, c.Bindings, appDefaultSpace)
+
+	err = applicationClient.MergeBindings(params.ApplicationMergeBindingsArgs{
+		Args: []params.ApplicationMergeBindings{
+			{
+				ApplicationTag: names.NewApplicationTag(c.ApplicationName).String(),
+				Bindings:       c.Bindings,
+				Force:          c.Force,
+			},
+		},
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Emit binding changelog after a successful call to MergeBindings.
+	for _, change := range bindingsChangelog {
+		ctx.Infof(change)
+	}
+	return nil
+}
+
+func (c *bindCommand) validateEndpointNames(newCharmEndpoints set.Strings, oldEndpointsMap, userBindings map[string]string) error {
+	for epName := range userBindings {
+		if _, exists := oldEndpointsMap[epName]; exists || epName == "" {
+			continue
+		}
+
+		if !newCharmEndpoints.Contains(epName) {
+			return errors.NotFoundf("endpoint %q", epName)
+		}
+	}
+	return nil
+}
+
+func (c *bindCommand) parseBindExpression(apiRoot base.APICallCloser) error {
+	if c.BindExpression == "" {
+		return nil
+	}
+
+	// Fetch known spaces from server
+	knownSpaceList, err := c.NewSpacesClient(apiRoot).ListSpaces()
```

Comment:
I thought commands stuck with 1 primary API.  If there was code needed from a different API, it was duplicated or moved to api.common to be shared.

Reply:
The deploy and upgrade-charcm commands (same package) also use multiple APIs so I wanted to be consistent.